app/models/petition.rb

```diff
@@ -729,7 +729,15 @@ def publish(time = Time.current)
 
     build_pe_number
 
-    update(state: OPEN_STATE, open_at: time, closed_at: closing_date(time))
+    if collect_signatures?
+      update(state: OPEN_STATE, open_at: time, closed_at: closing_date(time))
+    else
+      ## Need to go through the stages -> open -> closed, and set referred at. 
+      update(state: OPEN_STATE, open_at: time, closed_at: closing_date(time))
+      close!(Time.current)
+      update(referral_threshold_reached_at: Time.current)
+      refer_or_reject!(Time.current)
+    end
   end
```

Comment:
We can rewrite this block as:

```
transaction do
  update!(state: OPEN_STATE, open_at: time, closed_at: closing_date(time))

  unless collect_signatures?
    if at_threshold_for_referral?
      update!(referral_threshold_reached_at: time)
    end

    close!(time)
    refer_or_reject!(time)
  end
end
```

That way it'll rollback any changes if a validation fails.

Reply:
I think you also need to change `closing_date`:

```
def closing_date(time)
  closed_at || (collect_signatures? ? Site.closed_at_for_opening(time) || time)
end
```